### PR TITLE
Build system overhaul (including Tenacity's build system)

### PR DESCRIPTION
**Pull request by generic-pers0n**
Resolves: #43

Build system overhaul, including parts of Tenacity's build system.

Current status: **Buildable on Linux (tested on Arch and Pop!_OS)**. More details:

* Saucedacity is now reporting the correct version number.
* Building with default settings works.
* Compiler caching is working now (tested with ccache)

**Broken/Believed to be broken**:

  * Using Tenacity's vcpkg submodule doesn't work.
    * wxWidgets is not found.
    * Maybe update to upstream vcpkg?
  * Our CI setup needs to be updated (namely CodeQL).
  * **Windows does not build**

- [x] I made sure the code compiles on my machine*
- [x] I made sure that my contributes are licensed under GPL v2 or later.*
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"

\*Indicates required.

### CODE REVIEW

**Comment by generic-pers0n** [comment]
Note to self: **stop adding new libraries and/or source files**! It drags this PR even further than it has to (at least in my opinion).

**Comment by generic-pers0n** [comment]
# Checklist

## Building
- [ ] Windows building
- [x] macOS building
- [x] Linux building
- [ ] vcpkg works
  * vcpkg partially works on Linux. wxWidgets is a little flaky.

## Packaging
- [ ] Windows packaging
- [x] macOS packaging
- [ ] Linux packaging

## Other features
- [x] Compiler caching with ccache, sccache, and maybe others.

**Comment by clort81** [comment]
Hi! Building tenacity works fine here.  Saucedacity build borked on aarch64 devuan linux (ceres, similar to debian sid but without systemd)
```
wx-config --version
3.1.5  (that's my system wxwidgets, conan is pulling-in ~/.conan/data/wxwidgets/3.1.3.2-audacity

Conan version 1.53.0
Python 3.10.7
-- Conan: Adding audacity remote repository (https://artifactory.audacityteam.org/artifactory/api/conan/conan-local) verify ssl (True)
WARN: Remotes registry file missing, creating default one in /home/plab/.conan/remotes.json
-- The ASM_NASM compiler identification is NASM
-- Found assembler: /usr/bin/nasm
-- Build Info:
--   Host System: Linux-5.15.0+
--   Host System Name: Linux
--   Host System Processor: aarch64
--   Host System Version: 5.15.0+
--   Host System Architecture: 64-bit
--   Compiler: /usr/bin/c++
--   Compiler Version: 9.4.0
--   Compiler Standard: 17
--   Compiler Standard Required: ON
--   Compiler Extensions: 
--   Current Commit: 9d848204c

```
From CMAKE error log:
```
Building CXX object CMakeFiles/cmTC_12fdf.dir/src.cxx.o
/usr/bin/c++ -DHAVE_SSE  -O2 -ffast-math -ftree-vectorize -fno-strict-overflow -funsafe-math-optimizations -march=armv8-a+
crypto+crc+aes+sha2    -msse -std=gnu++17 -o CMakeFiles/cmTC_12fdf.dir/src.cxx.o -c /pr/Audio/saucedacity/build/CMakeFiles
/CMakeTmp/src.cxx
c++: error: unrecognized command line option '-msse'
```
Did you add sse and sse2 as a hard requirement?  Arm dinna have that!
There's an interesting project sse2neon that could maybe be used to generate alternative NEON vector code.  
Tenacity seems to build without sse/sse2 - can you test for it and not require if building on ARM?
[CMakeError.log.gz](https://github.com/tenacityteam/saucedacity/files/9950248/CMakeError.log.gz)

It's possible I misidentified the problem ofc. I am not an audacity/tenacity developer.

**Comment by generic-pers0n** [comment]
> Conan version 1.53.0
> Python 3.10.7
> -- Conan: Adding audacity remote repository (https://artifactory.audacityteam.org/artifactory/api/conan/conan-local) verify ssl (True)

Looks like this is with the old build Conan-enabled system. Have you tried this branch which does not use Conan?

> From CMAKE error log:
> 
> ```
> Building CXX object CMakeFiles/cmTC_12fdf.dir/src.cxx.o
> /usr/bin/c++ -DHAVE_SSE  -O2 -ffast-math -ftree-vectorize -fno-strict-overflow -funsafe-math-optimizations -march=armv8-a+
> crypto+crc+aes+sha2    -msse -std=gnu++17 -o CMakeFiles/cmTC_12fdf.dir/src.cxx.o -c /pr/Audio/saucedacity/build/CMakeFiles
> /CMakeTmp/src.cxx
> c++: error: unrecognized command line option '-msse'
> ```
> 
> Did you add sse and sse2 as a hard requirement? Arm dinna have that!

On x86 (32 or 64 bit), SSE and SSE2 are a hard requirement, **but not on ARM**. This is a bug.

What's interesting, however, is that I was able to build Saucedacity successfully on a Raspberry Pi 3 on Raspberry Pi OS.

> There's an interesting project sse2neon that could maybe be used to generate alternative NEON vector code. Tenacity seems to build without sse/sse2 - can you test for it and not require if building on ARM? [CMakeError.log.gz](https://github.com/tenacityteam/saucedacity/files/9950248/CMakeError.log.gz)

I have had plans to clean up Saucedacity's SIMD math functions, dropping SSE code, and adding support for newer intrinsics such as AVX on x86_64 and NEON on ARM (targeting aarch64). For the longest time, I've been waiting for this PR to be completed and merged among name changes, moving the repository to Codeberg, and getting other necessary maintenance tasks out of the way. No code for this plan exists yet, but I hope to commit to refactoring our SIMD math functions one of these days.

**Comment by generic-pers0n** [comment]
Ah, yet another thing I have adapted from Tenacity's build system: `RULE_LAUNCH_COMPILE` in CMake. According to the CMake documentation and [this CMake issue](https://gitlab.kitware.com/cmake/cmake/-/issues/22405) by Be-ing, this property is used internally for CTest.

Although I believe there is no harm necessarily done in using internal things in general, I also believe that we should avoid them as best as we can in general as well.  However, the primary motivation is to attempt to fix Windows builds since the failure resides in `saucedacity.rc.res.obj` not being found. At least, according to MSVC, it ignores an unknown option `/fo`, and at least we can maybe fix that if this change doesn't fix Windows builds.

**TL;DR**: We were using an internal property, and I hope fixing this issue fixes Windows builds, although I'm prepared to react appropriately if this does not.

**Comment by generic-pers0n** [comment]
Now the issue is focusing on ToolManager.cpp. I rebased this branch thinking things would work if I dropped the commit adding "<wx/eventfilter.h" to ToolManager.h. The build still failed on Windows but at CMake (or so it appeared) but also failed on Linux and macOS.😞

At the very least, we've pushed Windows builds further, and now we're getting a different error. However, this error was the same error that was already fixed some time ago prior to me reverting it, so we're back to a similar point in time...

**Comment by generic-pers0n** [comment]
Okay. At this point, I am able to verify that Windows building works with MSVC and Ninja **without compiler cache**. This will work for now, despite CI failing.

**I am now marking this PR as ready to merge, although I have a few little things to clean up and change before we are in an ideal state**.

**Comment by generic-pers0n** [comment]
Alright. I'm done working on this PR and I am requesting reviews on this PR. Windows CI is broken, but that doesn't mean that Saucedacity will not build on Windows at all. A solution is to not use sccache. Ninja and MSVC work according to my tests.

@caughtquick, if you have other stuff to add, go ahead and commit it. Don't forget to rebase your branch off the latest upstream if necessary.

When everything if entirely finalized and reviews are in, I will merge this PR, and we will finally be able to move on with other things.

**Comment by generic-pers0n** [comment]
Since work has been finished and everything is ready to merge, **I will give this PR by the end of next week for anyone to add anything, to object, etc**. Otherwise we can resolve these things later :)

**Comment by generic-pers0n** [comment]
Actually, I'll likely merge this around this weekend since it looks like everything's good so far. :D
